cmd/snap-bootstrap/triggerwatch/evdev.go

```diff
@@ -0,0 +1,242 @@
+	if err != 0 {
+		return false, err
+	}
+	byteIdx := e.keyCode / 8
+	keyMask := byte(1 << (e.keyCode % 8))
+	isDown := keyBitmap[byteIdx]&keyMask != 0
+	return isDown, nil
+}
+
+func (e *evdevKeyboardInputDevice) WaitForTrigger(ch chan keyEvent) {
+	logger.Noticef("%s: starting wait", e)
+
+	// XXX: do not mess with setting the key repeat rate, as it's cumbersome
+	// and golang-evdev SetRepeatRate() parameter order is actually reversed
+	// wrt. what the kernel does. The evdev interprets EVIOCSREP arguments
+	// as (delay, repeat)
+	// https://elixir.bootlin.com/linux/latest/source/drivers/input/evdev.c#L1072
+	// but the wrapper is passing is passing (repeat, delay)
+	// https://github.com/gvalkov/golang-evdev/blob/287e62b94bcb850ab42e711bd74b2875da83af2c/device.go#L226-L230
+
+	holdState, err := e.probeKeyState()
+	if err != nil {
+		ch <- keyEvent{Err: fmt.Errorf("cannot obtain initial key state: %v", err), Dev: e}
+	}
+	if holdState {
+		// looks like the key is pressed initially, we don't know when
+		// that happened, but prentend got pressed now
+		logger.Noticef("%s: key is already down", e)
+	}
+
+	type evdevEvent struct {
+		kev *evdev.KeyEvent
+		err error
+	}
+
+	// buffer large enough to collect some events
+	evChan := make(chan evdevEvent, 10)
+
+	monitorKey := func() {
+		for {
+			ies, err := e.dev.Read()
+			if err != nil {
+				evChan <- evdevEvent{err: err}
+				break
+			}
+			for _, ie := range ies {
+				if ie.Type != evdev.EV_KEY || ie.Code != e.keyCode {
+					continue
+				}
+				kev := evdev.NewKeyEvent(&ie)
+				evChan <- evdevEvent{kev: kev}
+			}
+		}
+		close(evChan)
+	}
+
+	go monitorKey()
+
+	holdTimer := time.NewTimer(holdToTrigger)
+	// no sense to keep it running later either
+	defer holdTimer.Stop()
+
+	if !holdState {
+		// key isn't held yet, stop the timer
+		holdTimer.Stop()
+	}
+Loop:
+	for {
+		select {
+		case ev := <-evChan:
+			if ev.err != nil {
+				holdTimer.Stop()
+				ch <- keyEvent{Err: err, Dev: e}
+				break Loop
+			}
+			kev := ev.kev
+			switch kev.State {
+			case evdev.KeyDown:
+				if holdState {
+					// unexpected, but possible if we missed
+					// a key up event right after checking
+					// the initial keyboard state when the
+					// key was still down
+					if !holdTimer.Stop() {
+						// drain the channel before the
+						// timer gets reset
+						<-holdTimer.C
+					}
+				}
+				holdState = true
+				// timer is stopped at this point
+				holdTimer.Reset(holdToTrigger)
+				logger.Noticef("%s: trigger key down", e)
+			case evdev.KeyHold:
+				if !holdState {
+					holdState = true
+					// timer is not running yet at this point
+					holdTimer.Reset(holdToTrigger)
+					logger.Noticef("%s: unexpected hold without down", e)
+				}
+			case evdev.KeyUp:
+				// no need to drain the channel, if it expired,
+				// we'll handle it in next iteration
```

Comment:
how do we know that this will be true and we won't get another evdev event on evChan first, as that's also a buffered channel?

Reply:
The events are serialized. There's only Up/Down/Hold support, in which case Up stops the timer, hold does nothing, down stops and restarts the timer. Repeated events of the same kind are fine too, in case of Up, the timer would get stopped each time, Hold distinguishes key states, and Down stops and resets the timer as needed.